src/jdk.compiler/share/classes/com/sun/tools/javac/main/JavaCompiler.java

```diff
@@ -977,6 +977,14 @@ public void compile(Collection<JavaFileObject> sourceFileObjects,
                 log.printVerbose("total", Long.toString(elapsed_msec));
             }
 
+            // In case an Abort was thrown before processAnnotations could be called,
+            // we could have deferred diagnostics that haven't been reported. At the
+            // time of this comment's writing, this can only ever occur through a
+            // missing java.lang error.
+            if (deferredDiagnosticHandler != null) {
+                deferredDiagnosticHandler.reportDeferredDiagnostics();
```

Comment:
The CI run shows some real test failures that still need to be addressed

```
java.lang.NullPointerException: Cannot invoke "java.util.Queue.poll()" because "this.deferred" is null
        at jdk.compiler/com.sun.tools.javac.util.Log$DeferredDiagnosticHandler.reportDeferredDiagnostics(Log.java:169)
        at jdk.compiler/com.sun.tools.javac.util.Log$DeferredDiagnosticHandler.reportDeferredDiagnostics(Log.java:163)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:985)
        at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:319)
        at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:178)
        at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:64)
        at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:50)
```

It looks like `reportDeferredDiagnostics` is now being called twice for some invocations, and when it is called it sets the queue of deferred diagnostics to `null`, so the second call fails: https://github.com/openjdk/jdk/blob/45414fc2dfa41cbbfc6de7fec15eb47f41cf8986/src/jdk.compiler/share/classes/com/sun/tools/javac/util/Log.java#L173

One option would be to set `deferredDiagnosticHandler` to `null` after calling `reportDeferredDiagnostics` in `JavaCompiler`, so we can keep track of whether the deferred diagnostics have already been reported

Reply:
> One option would be to set `deferredDiagnosticHandler` to `null` after calling `reportDeferredDiagnostics` in `JavaCompiler`, so we can keep track of whether the deferred diagnostics have already been reported

Paula had a better suggestion of moving the new logic to the `Abort` handler, which ensures it's only called once and fixes the test regressions.